celery_serverless/handler.py

```diff
@@ -82,10 +84,39 @@ def watchdog(event, context):
     else:
         watched = KombuQueueLengther(queue_url, 'celery')   # TODO: Allow queue name to be chosen
 
-    Watchdog(communicator=build_intercom(intercom_url), name=lock_name, lock=lock, watched=watched).monitor()
+    watchdog = Watchdog(communicator=build_intercom(intercom_url), name=lock_name, lock=lock, watched=watched)
+
+    try:
+        remaining_seconds = context.get_remaining_time_in_millis() / 1000.0
+    except Exception as e:
+        logger.exception('Could not got remaining_seconds. Is the context right?')
+        remaining_seconds = 5 * 60 # 5 minutes by default
+
+    spare_time = 30  # Should be enough time to retrigger this FaaS again
+    fulfilled = False
+    try:
+        with timeout_context(remaining_seconds-spare_time):
+            watchdog.monitor()
+            fulfilled = True
+    except TimeoutError:
+        logger.warning('Still stuff to monitor but this Function is out of time. Preparing to reinvoke the Watchdog')
+
+        # Be sure that the lock is released. Then reinvoke.
+        try:
+            lock.release()
+        except (RuntimeError, AttributeError):
+            pass
+        else:
+            time.sleep(1)  # Let distributed locks to propagate
```

Comment:
Should we worry about distributed right now?

Reply:
Is worth to not take a 1sec nap?